### PR TITLE
implement symbol-avoid-edges

### PR DESCRIPTION
**Pull request by ansis**
Prevent symbols from colliding with tile edges when `symbol-avoid-edges` is true.

ref https://github.com/mapbox/mapbox-gl-native/issues/3582

render tests: https://github.com/mapbox/mapbox-gl-test-suite/commit/bc0272cdde41d027a5df1ca8fb3aa1bc49aa8149

### CODE REVIEW

**Comment by ansis** [comment]
@lucaswoj can you review?

File: js/symbol/collision_tile.js [hunk 1]
```diff
@@ -52,61 +66,95 @@ CollisionTile.prototype.placeCollisionFeature = function(collisionFeature) {
...
-                // so unblock the new box at the scale that the old box disappears.
-                collisionFreeScale = blocking.maxScale;
-            }
 
-            if (collisionFreeScale > box.maxScale) {
-                // If the box can only be shown after it is visible, then the box can never be shown.
-                // But the label can be shown after this box is not visible.
-                collisionFreeScale = box.maxScale;
-            }
+CollisionTile.prototype.checkForCollision = function(minPlacementScale, anchorPoint, box, blockingAnchorPoint, blocking) {
```
Comment (lucaswoj) [inline js/symbol/collision_tile.js]: Would it make more sense for this method to be called `findMinPlacementScale`?

Comment (lucaswoj) [inline js/symbol/collision_tile.js]: I suppose keeping these method names as similar as possible to those in native is most important.

File: js/symbol/collision_tile.js [hunk 2]
```diff
@@ -52,61 +66,95 @@ CollisionTile.prototype.placeCollisionFeature = function(collisionFeature) {
...
 
-            if (collisionFreeScale > minPlacementScale &&
-                    collisionFreeScale >= blocking.placementScale) {
-                // If this collision occurs at a lower scale than previously found collisions
-                // and the collision occurs while the other label is visible
+    // Find the lowest scale at which the two boxes can fit side by side without overlapping.
```
Comment (lucaswoj) [inline js/symbol/collision_tile.js]: A clarifying comment about what exactly is meant by `scale` would go a long way here.

File: js/symbol/collision_tile.js [hunk 3]
```diff
@@ -28,6 +31,17 @@ function CollisionTile(angle, pitch) {
     // The amount the map is squished depends on the y position.
     // Sort of account for this by making all boxes a bit bigger.
     this.yStretch = Math.pow(this.yStretch, 1.3);
+
+    this.edges = [
```
Comment (lucaswoj) [inline js/symbol/collision_tile.js]: Do we need to create an instance of `edges` per tile? Couldn't there just be one global instance?

Comment (lucaswoj) [inline js/symbol/collision_tile.js]: Ah. They might have different extents. Never mind.

**Comment by lucaswoj** [comment]
Looks :+1: overall. Ready to ship once minor comments are addressed.

**Pull request merged**